go/vt/vtctl/reparentutil/durability_policies.go

```diff
@@ -0,0 +1,238 @@
+	newDurabilityCreationFunc, found := durabilityPolicies[name]
+	if !found {
+		return fmt.Errorf("durability policy %v not found", name)
+	}
+	log.Infof("Setting durability policy to %v", name)
+	curDurabilityPolicyMutex.Lock()
+	defer curDurabilityPolicyMutex.Unlock()
+	curDurabilityPolicy = newDurabilityCreationFunc(durabilityParams)
+	return nil
+}
+
+// PromotionRule returns the promotion rule for the instance.
+func PromotionRule(tablet *topodatapb.Tablet) promotionrule.CandidatePromotionRule {
+	curDurabilityPolicyMutex.Lock()
+	defer curDurabilityPolicyMutex.Unlock()
+	return curDurabilityPolicy.promotionRule(tablet)
+}
+
+// SemiSyncAckers returns the primary semi-sync setting for the instance.
+// 0 means none. Non-zero specifies the number of required ackers.
+func SemiSyncAckers(tablet *topodatapb.Tablet) int {
+	curDurabilityPolicyMutex.Lock()
+	defer curDurabilityPolicyMutex.Unlock()
+	return curDurabilityPolicy.semiSyncAckers(tablet)
+}
+
+// IsReplicaSemiSync returns the replica semi-sync setting from the tablet record.
+// Prefer using this function if tablet record is available.
+func IsReplicaSemiSync(primary, replica *topodatapb.Tablet) bool {
+	curDurabilityPolicyMutex.Lock()
+	defer curDurabilityPolicyMutex.Unlock()
+	return curDurabilityPolicy.isReplicaSemiSync(primary, replica)
+}
+
+//=======================================================================
+
+// durabilityNone has no semi-sync and returns NeutralPromoteRule for Primary and Replica tablet types, MustNotPromoteRule for everything else
+type durabilityNone struct{}
+
+func (d *durabilityNone) promotionRule(tablet *topodatapb.Tablet) promotionrule.CandidatePromotionRule {
+	switch tablet.Type {
+	case topodatapb.TabletType_PRIMARY, topodatapb.TabletType_REPLICA:
+		return promotionrule.Neutral
+	}
+	return promotionrule.MustNot
+}
+
+func (d *durabilityNone) semiSyncAckers(tablet *topodatapb.Tablet) int {
+	return 0
+}
+
+func (d *durabilityNone) isReplicaSemiSync(primary, replica *topodatapb.Tablet) bool {
+	return false
+}
+
+//=======================================================================
+
+// durabilitySemiSync has 1 semi-sync setup. It only allows Primary and Replica type servers to acknowledge semi sync
+// It returns NeutralPromoteRule for Primary and Replica tablet types, MustNotPromoteRule for everything else
+type durabilitySemiSync struct{}
+
+func (d *durabilitySemiSync) promotionRule(tablet *topodatapb.Tablet) promotionrule.CandidatePromotionRule {
+	switch tablet.Type {
+	case topodatapb.TabletType_PRIMARY, topodatapb.TabletType_REPLICA:
+		return promotionrule.Neutral
+	}
+	return promotionrule.MustNot
+}
+
+func (d *durabilitySemiSync) semiSyncAckers(tablet *topodatapb.Tablet) int {
+	return 1
+}
+
+func (d *durabilitySemiSync) isReplicaSemiSync(primary, replica *topodatapb.Tablet) bool {
+	switch replica.Type {
+	case topodatapb.TabletType_PRIMARY, topodatapb.TabletType_REPLICA:
```

Comment:
Why would you ever pass a PRIMARY as `replica` to this function?

Reply:
There are two reasons why that would happen - 
1. As a fail safe for, if we have multiple tablets thinking themselves to be primary
2. We want to know whether the current primary will be a sending semi sync acks for a replica (which is a potential primary at this point). So the `replica` argument can be a tablet of type Primary.